cmake/modules/AddSwift.cmake

```diff
@@ -434,18 +438,35 @@ function(_add_variant_link_flags)
     list(APPEND link_libraries "bsd" "atomic")
     list(APPEND result "-Wl,-Bsymbolic")
   elseif("${LFLAGS_SDK}" STREQUAL "ANDROID")
-    list(APPEND link_libraries "dl" "log" "atomic" "icudataswift" "icui18nswift" "icuucswift")
+    if("${SWIFT_HOST_VARIANT}" STREQUAL "android")
+      list(APPEND link_libraries "dl" "log" "atomic" "icudata" "icui18n" "icuuc")
```

Comment:
This is wrong; you **MUST** use the swift variant libraries as the versions on Android are not guaranteed to meet the requirements.

Reply:
Agree. IIRC those libraries are distributed with Android, but they are not part of the NDK, so they should not be considered public. The problem with ICU is that the version number is baked into the symbols, so if you compile in your machine, it might be version 59, but someone tries to use it one year from now, and the package will not work, because their ICU libraries provide version 61 symbols. That's why Swift decides to ship their own ICU (it needs the different name to avoid loading the system libraries).

Reply:
>That's why Swift decides to ship their own ICU

Yes, I ran into this early on, maybe before I installed the libicu-dev Termux package mentioned above, but CMake now automatically finds and links libswiftCore.so against that local ICU 63.1 package, so this should be fine.

Reply:
There is no guarantee about the version being shipped, so, it should always use the locally built version.

Reply:
It is, the ICU version that ships with Termux's libicu package. Since Termux comes with its own ICU with a known version, that's what it's using. It isn't using the Android system ICU in /system/lib64, which can vary based on which Android version is installed.

Reply:
>That explains things. If you plan to distribute this, however, you are forced to fix onto a ICU version package.

Yep.

>If a 3rd party users upgrades ICU, but not the Swift package, things will break anyway.

Sure, but Termux can always bump the Swift package if that happens, just as [they recently bumped a bunch of packages](https://github.com/termux/termux-packages/commit/5871ed7b26b35f12f75a8aa0e5643dc03387bfae) when they upgraded all the libraries they pull from the NDK to the latest.

This is a known and easily solved problem, and no different than how [the official build of Swift for Ubuntu 18.04](https://swift.org/download/#releases) links against Ubuntu's system ICU 60 package, rather than shipping its own ICU.

>I also don’t know how Termux sets things up, but IIRC this line will not link to the full path of the libraries, so they will be searched in the runtime search path, and the system ones might be the first one to be loaded (because their name is compatible with Termux name).

Termux adds the its lib path to LD_LIBRARY_PATH, so it will always find the Termux ICU package first:
```
termux$ env | grep LD_LIBRARY_PATH
LD_LIBRARY_PATH=/data/data/com.termux/files/usr/lib
```
Even if it didn't, the FindICU CMake config finds the local Termux ICU libraries and links against those, which are symbolic links to the versioned libraries like libicuuc.so.63, so the linker adds those versioned library names to the list too.

Searching the test results now, one does fail when it resets LD_LIBRARY_PATH, because it doesn't find the versioned ICU library:
```
: 'RUN: at line 65';   env LD_LIBRARY_PATH=/abc/ '/data/data/com.termux/files/home/src/swift/build/swift/bin/swift' -### -target x86_64-unknown-linux-gnu -L/foo/ -L/bar/ /data/data/com.termux/files/home/src/swift/swift/test/Driver/options-interpreter.swift | /data/data/com.termux/files/home/src/swift/swift/utils/PathSanitizingFileCheck --sanitize 'BUILD_DIR=/data/data/com.termux/files/home/src/swift/build/swift' --sanitize 'SOURCE_DIR=/data/data/com.termux/files/home/src/swift/swift' --use-filecheck '/data/data/com.termux/files/home/src/swift/build/llvm/bin/FileCheck' -check-prefix=CHECK-LINUX-COMPLEX${LD_LIBRARY_PATH+_LAX} /data/data/com.termux/files/home/src/swift/swift/test/Driver/options-interpreter.swift
--
Exit Code: 2

Command Output (stderr):
--
CANNOT LINK EXECUTABLE "/data/data/com.termux/files/home/src/swift/build/swift/bin/swift": library "libicudata.so.63" not found
```
Not a big deal, you just can't reset the LD_LIBRARY_PATH like that in Termux.

>You had a phrase about some test failing, I wonder if those are the ones that tries to use ICU symbols (I think there are some in the stdlib, and many more in Foundation).

I would imagine many more tests fail with informative errors like the one above if that were the case, but a grep for icu only turned up that one. I suppose it's possible that the validation test suite doesn't really check calling ICU functions anywhere, but that's not possible, is it?

Again, you're putting a lot of thought into a problem that the Termux devs already deal with regularly, and is no different than how Swift already packages its linux release.